### PR TITLE
Updating Kubernetes tests to properly test missing endpoints code path

### LINKED ISSUE #1307: Update Kubernetes tests to properly handle endpoints not existing
In #1295 I introduced a temporary work around, this issue is for removing that and cleaning up the tests that break.

The code that originally causes the issue is this:
```
func (c clientMock) GetEndpoints(namespace, name string) (*v1.Endpoints, bool, error) {
	if c.apiEndpointsError != nil {
		return &v1.Endpoints{}, false, c.apiEndpointsError
	}

	for _, endpoints := range c.endpoints {
		if endpoints.Namespace == namespace && endpoints.Name == name {
			return endpoints, true, nil
		}
	}

	return &v1.Endpoints{}, true, nil
}
```

The final return has true as the second value being returned even if no endpoints for the service were found. This means the tests never hit the code path for `!exists` here: https://github.com/containous/traefik/blob/master/provider/kubernetes.go#L240

Instead what happens is all the tests that don't specify endpoints go through and hit the code path for `len(endpoints.Subsets) == 0`.

We need to update the mock `GetEndpoints` function to return `false` for the second return value when no endpoints are found, fix all the tests that break, and then ensure that there is a test that tests the current code path being used.

Here are the tests that fail when the `GetEndpoints` function returns correctly: TestHostlessIngress, TestLoadMultipleNamespacedIngresses, TestLoadNamespacedIngresses, TestGetPassHostHeader, TestOnlyReferencesServicesFromOwnNamespace, TestLoadIngresses

Related to this is defining and possibly changing what happens when endpoints are not found. Currently when a service is not found the referencing frontend is removed. However if no endpoints are found we still keep the frontend. Should we be removing the frontend if no endpoints are found?

### PR DESCRIPTION
**Pull request by Regner**
This fixes #1307

### CODE REVIEW

**Comment by Regner** [comment]
As a heads up when I run the provider tests locally I get the following failed test:
```
--- FAIL: TestDockerLoadDockerServiceConfig (0.00s)
	docker_test.go:3124: expected map[string]*types.Backend{"backend-foo-service":(*types.Backend)(0xc42062cc90)}, got map[string]*types.Backend{"backend-foo":(*types.Backend)(0xc4206576b0)}
```

This PR doesn't touch anything relating to that test. Curious if it fails in Travis. Haven't looked into why but I am running from head.

File: provider/kubernetes_test.go [hunk 1]
```diff
@@ -499,12 +495,7 @@ func TestGetPassHostHeader(t *testing.T) {
 	expected := &types.Configuration{
 		Backends: map[string]*types.Backend{
 			"foo/bar": {
-				Servers: map[string]types.Server{
-					"1": {
-						URL:    "http://10.0.0.1:801",
-						Weight: 1,
-					},
-				},
+				Servers:        map[string]types.Server{},
```
Comment (timoreimann) [inline provider/kubernetes_test.go]: What's the reason we are able to remove / empty a bunch of server definitions?

Comment (timoreimann) [inline provider/kubernetes_test.go]: Oh of course, it's because we now model missingness correctly, right?

Comment (Regner) [inline provider/kubernetes_test.go]: Yea, see this line: https://github.com/containous/traefik/pull/1436/files#diff-4c0e9bf481017297ab14b697157b2675R2079

**Comment by Regner** [comment]
One of the annoying things was at first I just left the `Servers:        map[string]types.Server{},` line out, and the tests failed but looked 100% identical according to the output but were actually different.

**Comment by timoreimann** [comment]
@Regner :+1:

In the referencing issue, you also mention the fact that we keep frontends dangling when no Endpoints can be found. Do you know what the effect of that is? And should we address it in this PR or a different one? (I'm open to both paths.)

**Comment by Regner** [comment]
I don't know what the effect of that is and am unsure how that should be handled. I think it needs some discussion first so should probably be a different issue and PR.

**Comment by timoreimann** [comment]
@Regner makes sense to me.

**Comment by timoreimann** [comment]
@Regner could you rebase and fix the conflicts? We recently moved all providers into separate sub-packages. Sorry for the inconveniences!

**Pull request merged**